csrc/scheduler/vectorize_helper.cpp

```diff
@@ -64,63 +64,60 @@ ContiguousInnerDimensionsMapper::ContiguousInnerDimensionsMapper(
-      "Reference: ",
-      reference->toString());
+  // Exclude reduction IDs if the reference is a fusion input as they
+  // don't manifest at all in the fusion. This simplifies the
+  // analysis in getContigMergeOfInnerSize, which only looks at
+  // non-reduction rfactor domains. Including reduction domains here
+  // can result in incorrect ordering
+  // NOTE: this is necessary to enable vectorization in
+  // NVFuserTest.FusionSegmentReduceSoftmax_CUDA
+  auto rfactor_domain = reference->getMaybeRFactorDomain();
+  auto filtered_ids = ids;
+  if (reference->isFusionInput()) {
+    auto remove_reduction = [](auto& container) {
+      container.erase(
+          std::remove_if(
+              container.begin(),
+              container.end(),
+              [](IterDomain* id) { return id->isReduction(); }),
+          container.end());
+    };
+    remove_reduction(rfactor_domain);
+    remove_reduction(filtered_ids);
+  }
 
   // Record while processing reference's information
   recording_ = true;
-  std::shared_ptr<Information> reference_information;
+  for (auto id : filtered_ids) {
+    addProjectedExtent(id, commonOrConstExtent(ca_map_, id));
+  }
 
   // Ordering of dimensions is important in this analysis, if an ordering is
   // contiguous in the reference, but not the target tensor views, then we
   // cannot consider that a contiguous merge dimension for vectorization.
-  std::vector<IterDomain*> reordered_rfactor;
-  for (auto id : reference->getMaybeRFactorDomain()) {
-    // Exclude reduction IDs if the reference is a fusion input as they
-    // don't manifest at all in the fusion. This simplifies the
-    // analysis in getContigMergeOfInnerSize, which only looks at
-    // non-reduction rfactor domains. Including reduction domains here
-    // can result in incorrect ordering
-    if (reference->isFusionInput() && id->isReduction()) {
-      continue;
-    }
-    // If not a fusion input, can this be a reduction domain?
-    TORCH_INTERNAL_ASSERT(
-        !id->isReduction(),
-        "Unexpected reduction domain: ",
-        id->toString(),
-        " of tensor ",
-        reference->toString());
-    if (std::find(reference_ids.begin(), reference_ids.end(), id) !=
-        reference_ids.end()) {
-      reordered_rfactor.push_back(id);
-      // Initiailze the extent for the mapped iter domain
-      addProjectedExtent(id, commonOrConstExtent(ca_map_, id));
-    } else if (!id->isBroadcast()) {
-      // Ignore broadcasts in the reference. Otherwise, remove non-contiguous
-      // IDs in the reference tensor as this is the contiguous mapper.
-      reordered_rfactor.clear();
+  auto projected_rfactor = projectId(filtered_ids, rfactor_domain);
+
+  // Need to somehow make sure projected_rfactor is indeed on the reference
+  // tensor, otherwise, getProjectedExtent seems to be complaining. I'm
+  // wondering if we should instead canonicalize the key for projected extent
```

Comment:
I think `projectId` should always return a vector whose members are in the given `to` set, for this case `projectId` must return IDs that is in the rFactor domain. Otherwise this is a bug in `projectId`. Could you provide more information when this fail to be true?

Reply:
I was hoping that, but it is not what I'm seeing.

I vaguely remember that previously I was trying to project from reference1's leaf domain to reference2's rfactor domain and that's where I found the output doesn't always return IterDomain in the `to` set. So at least that part is no longer needed.

Let me try to revert this and see if any other failure would pop up.

Reply:
Looks like Xiang is right here. I'm reverting this one.